charon/src/register.rs

```diff
@@ -1,130 +1,260 @@
-        };
+fn get_opaque_declaration(
+    tcx: TyCtxt,
+    id: DefId,
+    kind: DeclKind,
+    name: &Name,
+) -> Option<Declaration> {
+    trace!("opaque declaration {}", name);
+
+    if is_primitive_decl(kind, id, name) {
+        return None;
+    }
+
+    // TODO: we check this here and in translate_types
+    check_decl_generics(kind, tcx, id);
+    Some(Declaration::new_opaque(id, kind))
+}
+
+struct RegisterContext<'a, 'b, 'c> {
+    rustc: TyCtxt<'a>,
+    sess: &'b Session,
+    crate_info: &'c CrateInfo,
+}
+
+pub type RegisteredDeclarations = LinkedHashMap<DefId, Declaration>;
+
+/// Structure used to register declarations : see
+/// [DeclarationsRegister::register_opaque_declaration] and
+/// [DeclarationsRegister::register_visible_declaration].
+struct DeclarationsRegister {
+    decl_ids: LinkedHashSet<DefId>,
+    decls: RegisteredDeclarations,
+}
+
+impl DeclarationsRegister {
+    fn new() -> DeclarationsRegister {
+        DeclarationsRegister {
+            decl_ids: LinkedHashSet::new(),
+            decls: RegisteredDeclarations::new(),
+        }
+    }
+
+    /// Indicates if the declaration is being (or has been) added.
+    fn knows(&self, id: &DefId) -> bool {
+        self.decl_ids.contains(id)
+    }
+
+    /// The declarations are registered in two steps :
+    /// When their exploration begin, we must first add their id.
+    /// This is done to prevent cycles while registering declarations.
+    ///
+    /// Should not be called outside [DeclarationsRegister]'s methods :
+    /// Use either [DeclarationsRegister::register_opaque_declaration]
+    /// or [DeclarationsRegister::register_visible_declaration] instead.
+    fn add_begin(&mut self, id: DefId) {
+        assert!(self.decl_ids.insert(id), "Already knows {:?}", id);
+    }
+
+    /// Finalize the registering of a declaration.
+    ///
+    /// Should not be called outside [DeclarationsRegister]'s methods :
+    /// Use either [DeclarationsRegister::register_opaque_declaration]
+    /// or [DeclarationsRegister::register_visible_declaration] instead.
+    fn add_end(&mut self, decl: Declaration) {
+        let id = decl.id;
+        assert!(self.knows(&id));
+
+        assert!(
+            self.decls.insert(id, decl).is_none(),
+            "Registered the same declaration {:?} twice",
+            id
+        );
+    }
+
+    /// Does not explore further the declaration content :
+    /// If the declaration was unknown, registers it without dependencies.
+    fn register_opaque_declaration(&mut self, tcx: TyCtxt, id: DefId, kind: DeclKind, name: &Name) {
+        if self.knows(&id) {
+            return;
+        }
+
+        get_opaque_declaration(tcx, id, kind, name).map(|decl| {
+            self.add_begin(id);
+            self.add_end(decl);
+        });
+    }
+
+    /// Registers a declaration and its dependencies recursively.
+    /// Only works on local declarations for now.
+    /// The visitor takes `&mut self` to avoid a double borrow.
```

Comment:
The method borrows `&mut self` from beginning to the end of the call, so `self` is not accessible in the visitor.
However, we want to access it inside the callback (to e.g. call this method on `self` again), that's why the borrow is given in the visitor.
So by "borrowing twice" I only meant that using `self` in the visitor is forbidden because it's already borrowed by the method.

Reply:
Ok!